src/query/storage/index.go

```diff
@@ -90,16 +98,48 @@ func FetchOptionsToM3Options(fetchOptions *FetchOptions, fetchQuery *FetchQuery)
 	}
 }
 
+var (
+	// byte representation for [1,2,3,4]
+	lookup = [4]byte{49, 50, 51, 52}
+)
+
+func queryKey(m models.Matchers) []byte {
+	l := len(m)
+	for _, t := range m {
+		l += len(t.Name) + len(t.Value)
+	}
+
+	key := make([]byte, l)
+	idx := 0
+	for _, t := range m {
+		idx += copy(key[idx:], t.Name)
+		key[idx] = lookup[t.Type]
+		idx += copy(key[idx+1:], t.Value)
+	}
+
+	return key
+}
+
 // FetchQueryToM3Query converts an m3coordinator fetch query to an M3 query
-func FetchQueryToM3Query(fetchQuery *FetchQuery) (index.Query, error) {
+func FetchQueryToM3Query(fetchQuery *FetchQuery, cache *QueryConvserionCache) (index.Query, error) {
 	matchers := fetchQuery.TagMatchers
+	k := queryKey(matchers)
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if val, ok := cache.LRU.Get(string(k)); ok {
+		return index.Query{Query: val}, nil
+	}
+
 	// Optimization for single matcher case.
 	if len(matchers) == 1 {
 		q, err := matcherToQuery(matchers[0])
 		if err != nil {
 			return index.Query{}, err
 		}
 
+		cache.LRU.Add(string(k), q)
```

Comment:
nit: rather than calling `.LRU.Add()`, can you put a thin wrapper `Add` function onto the query conversion cache?

Reply:
Any specific reason to do this?

Reply:
Reads better, more clarity; also if at some point we want to add more logic to the add function we can keep it in one place